cranelift/unwinder/src/stackwalk.rs

```diff
@@ -0,0 +1,230 @@
+//! exit trampoline. That exit trampoline is able to provide the
+//! "start FP" (FP at exit trampoline) and "end FP" (FP at entry
+//! trampoline) and this stack-walker can visit all Cranelift frames
+//! active on the stack between those two.
+//!
+//! This module provides a visitor interface to frames, but is
+//! agnostic to the desired use-case or consumer of the frames, and to
+//! the overall runtime structure that lets us find the first and last
+//! frame pointer of a series of Cranelift frames on the stack.
+
+use core::ops::ControlFlow;
+
+#[cfg(has_host_compiler_backend)]
+use crate::arch;
+
+/// Implementation necessary to unwind the stack, used by `Backtrace`.
+pub unsafe trait Unwind {
+    /// Returns the offset, from the current frame pointer, of where to get to
+    /// the previous frame pointer on the stack.
+    fn next_older_fp_from_fp_offset(&self) -> usize;
+
+    /// Returns the offset, from the current frame pointer, of the
+    /// stack pointer of the next older frame.
+    fn next_older_sp_from_fp_offset(&self) -> usize;
+
+    /// Load the return address of a frame given the frame pointer for that
+    /// frame.
+    unsafe fn get_next_older_pc_from_fp(&self, fp: usize) -> usize;
+
+    /// Debug assertion that the frame pointer is aligned.
+    fn assert_fp_is_aligned(&self, fp: usize);
+}
+
+/// A host-backed implementation of unwinding, using the native platform ABI
+/// that Cranelift has.
+#[cfg(has_host_compiler_backend)]
+pub struct UnwindHost;
+
+#[cfg(has_host_compiler_backend)]
+unsafe impl Unwind for UnwindHost {
+    fn next_older_fp_from_fp_offset(&self) -> usize {
+        arch::NEXT_OLDER_FP_FROM_FP_OFFSET
+    }
+    fn next_older_sp_from_fp_offset(&self) -> usize {
+        arch::NEXT_OLDER_SP_FROM_FP_OFFSET
+    }
+    unsafe fn get_next_older_pc_from_fp(&self, fp: usize) -> usize {
+        arch::get_next_older_pc_from_fp(fp)
+    }
+    fn assert_fp_is_aligned(&self, fp: usize) {
+        arch::assert_fp_is_aligned(fp)
+    }
+}
+
+/// An implementation specifically designed for unwinding Pulley's runtime stack
+/// (which might not match the native host).
+pub struct UnwindPulley;
+
+unsafe impl Unwind for UnwindPulley {
+    fn next_older_fp_from_fp_offset(&self) -> usize {
+        0
+    }
+    fn next_older_sp_from_fp_offset(&self) -> usize {
+        if cfg!(target_pointer_width = "32") {
+            8
+        } else {
+            16
+        }
+    }
+    unsafe fn get_next_older_pc_from_fp(&self, fp: usize) -> usize {
+        // The calling convention always pushes the return pointer (aka the PC
+        // of the next older frame) just before this frame.
+        *(fp as *mut usize).offset(1)
+    }
+    fn assert_fp_is_aligned(&self, fp: usize) {
+        let expected = if cfg!(target_pointer_width = "32") {
+            8
+        } else {
+            16
+        };
+        assert_eq!(fp % expected, 0, "stack should always be aligned");
+    }
+}
```

Comment:
Not necessarily required for this PR, but could this be moved back to Wasmtime? It feels a bit odd to define it all the way over here when Pulley is defined somewhere else.

Placement-wise the only use of `UnwindPulley` is [here](https://github.com/bytecodealliance/wasmtime/blob/5603ee7b099c837aa7194f1052e7c73cf4c36750/crates/wasmtime/src/runtime/store.rs#L1854) so it might make sense to add an `unwinder` method on `struct Interpreter` which would enable moving all of this to [this file](https://github.com/bytecodealliance/wasmtime/blob/main/crates/wasmtime/src/runtime/vm/interpreter.rs) (and a stub in [this file](https://github.com/bytecodealliance/wasmtime/blob/5603ee7b099c837aa7194f1052e7c73cf4c36750/crates/wasmtime/src/runtime/vm/interpreter_disabled.rs))

Reply:
Done!